File: src/algorithms/01_sorting/02_bubble_sort/implementations/ts/bubble_sort.ts
export const bubble_sort = (unsorted_array: Array<number>): Array<number> => {
  const cloned_array = [...unsorted_array]

  // Iterate array repeatedly until sorted
  while (true) {
    let swapped = false;

    // Iterate for window right item
    for (let i = 0; i < cloned_array.length - 1; i++) {
      if (cloned_array[i] > cloned_array[i + 1]) {
        [cloned_array[i], cloned_array[i + 1]] = [cloned_array[i + 1], cloned_array[i]];

        swapped = true;
      }
    }

    // If no item swapped, array is ordered
    if (!swapped) break;
  }

  return cloned_array;
};

File: src/algorithms/01_sorting/02_bubble_sort/implementations/rs/bubble_sort.rs
pub fn bubble_sort(unsorted_array: &Vec<i32>) -> Vec<i32> {
  let mut cloned_array = unsorted_array.clone();

  loop {
    let mut swapped: bool = false;

    for i in 0..cloned_array.len() - 1 {
      if cloned_array[i] > cloned_array[i + 1] {
        cloned_array.swap(i, i + 1);

        swapped = true
      }
    }

    if !swapped {
      break;
    }
  }

  cloned_array
}
